modules/core/04-channel/keeper/upgrade.go

```diff
@@ -668,23 +667,48 @@ func (k Keeper) ChanUpgradeTimeout(
 		)
 	}
 
+	// proof height must be from a height after timeout has elapsed.
+	if !upgrade.Timeout.Height.IsZero() && proofHeight.LT(upgrade.Timeout.Height) {
+		return errorsmod.Wrap(types.ErrInvalidUpgradeTimeout, "timeout height is not valid")
+	}
+
 	// proof must be from a height after timeout has elapsed. Either timeoutHeight or timeoutTimestamp must be defined.
 	// if timeoutHeight is defined and proof is from before timeout height, abort transaction
 	proofTimestamp, err := k.connectionKeeper.GetTimestampAtHeight(ctx, connection, proofHeight)
 	if err != nil {
 		return err
 	}
 
-	timeout := upgrade.Timeout
-	proofHeightIsInvalid := timeout.Height.IsZero() || proofHeight.LT(timeout.Height)
-	proofTimestampIsInvalid := timeout.Timestamp == 0 || proofTimestamp < timeout.Timestamp
-	if proofHeightIsInvalid && proofTimestampIsInvalid {
-		return errorsmod.Wrap(types.ErrInvalidUpgradeTimeout, "timeout has not yet passed on counterparty chain")
+	// if timeout timestamp is defined then the consensus time
+	// from proof height must be greater than timeout timestamp.
+	if upgrade.Timeout.Timestamp != 0 && proofTimestamp < upgrade.Timeout.Timestamp {
+		return errorsmod.Wrap(types.ErrInvalidUpgradeTimeout, "timeout timestamp is not valid")
```

Comment:
Isn't this essentially the same check as timeout packet? could we combine both height and timestamp checks into the same conditional like its done there:

```go
// check that timeout height or timeout timestamp has passed on the other end
proofTimestamp, err := k.connectionKeeper.GetTimestampAtHeight(ctx, connectionEnd, proofHeight)
if err != nil {
    return err
}

timeoutHeight := packet.GetTimeoutHeight()
if (timeoutHeight.IsZero() || proofHeight.LT(timeoutHeight)) &&
  (packet.GetTimeoutTimestamp() == 0 || proofTimestamp < packet.GetTimeoutTimestamp()) {
    return errorsmod.Wrap(types.ErrPacketTimeout, "packet timeout has not been reached for height or timestamp")
}
```

Reply:
yep it looks to be, we arrived at this two step approach to make things as readable as possible, but if we already have this condition I'm happy to take it as is

Reply:
actually it looks like the check you linked is the opposite. An error occurs if a timeout has **not** been reached. Here we are returning an error if a timeout **has** been reached.